ui/vty-lib/src/Ariadne/UI/Vty/Widget.hs

```diff
@@ -393,21 +423,27 @@ withWidgetState action = do
 -- Widget rendering
 ----------------------------------------------------------------------------
 
-drawWidget :: WidgetName -> p -> Widget p -> B.Widget WidgetName
+drawWidget :: WidgetName -> p -> Widget p -> WidgetDrawing
 drawWidget focus parent (Widget widget@WidgetInfo{..}) =
-    visible $ runReader (runReaderT (widgetDrawWithFocus focus widgetState) widget) parent
+    map visible $ runReader (runReaderT (widgetDrawWithFocus focus widgetState) widget) parent
   where
     visible w
       | focus == widgetName
       , not widgetIgnoreVisibility = B.visible w
       | otherwise = w
 
-drawWidgetChild :: WidgetName -> WidgetInfo s p -> WidgetNamePart -> B.Widget WidgetName
+drawWidgetChild :: WidgetName -> WidgetInfo s p -> WidgetNamePart -> WidgetDrawing
 drawWidgetChild focus widget@WidgetInfo{..} name =
-    maybe B.emptyWidget (drawWidget focus widget . withIgnoreVisibility) $ Map.lookup name widgetChildren
+    singleDrawing . maybe B.emptyWidget (last . drawWidget focus widget . withIgnoreVisibility) $ Map.lookup name widgetChildren
   where
     withIgnoreVisibility (Widget child) = Widget child{ widgetIgnoreVisibility = widgetIgnoreVisibility }
 
+singleDrawing :: B.Widget WidgetName -> WidgetDrawing
+singleDrawing = (:| [])
+
+layeredDrawing :: [B.Widget WidgetName] -> WidgetDrawing
```

Comment:
I don't like this function, because it's partial. AFAIU, it has two goals:
1. Somewhat encapsulate type of `WidgetDrawing`, so that if it's changed, you'll probably be able to change only definition of this body so that it still accepts a list instead of modifying all places where it's called.
2. You don't need to import `Data.List.NonEmpty` where you want to do layered drawing.

I manually checked that in all places it's used with a fixed list consisting of two items, so all usages are safe.
However, to make it more explicit and checked by compiler I propose to change type of this function to `B.Widget WidgetName -> [B.Widget WidgetName] -> WidgetDrawing`. This way both goals will still be achieved and at the same time this function will never `error`.

Reply:
You are right about both goals, I will change the `layeredDrawing` function as you say.